os/pm/pm_timedsuspend.c

```diff
@@ -106,51 +103,64 @@ static void timer_timeout(int argc, int domain_id, uint32_t hash_pid)
  *
  * Return Value:
  *   0 - success
- *   -1 - error
+ *  -1 - error
  *
  ************************************************************************/
 
 int pm_timedsuspend(int domain_id, unsigned int milliseconds)
 {
-	int hash_pid = PIDHASH(getpid());
-
-	if (domain_id >= CONFIG_PM_NDOMAINS || pm_domain_map[domain_id]==NULL) {
+	irqstate_t flags;
+	WDOG_ID wdog;
+	int tick_remain;
+	int ret = ERROR;
+	int delay = MSEC2TICK(milliseconds);
+	if ((domain_id < 0) || (domain_id >= CONFIG_PM_NDOMAINS) || (pm_domain_map[domain_id] == NULL)) {
 		set_errno(EINVAL);
-		return ERROR;
+		return ret;
 	}
-
-	/* Check if there is already a wdog lock timer running for 
-	 * the process */
-	if (pid_timer_map[hash_pid] != NULL) {
-		pmdbg("There is already a lock timer running for this process pid %d with domain %s\n", getpid(), pm_domain_map[domain_id]);
-		return ERROR;
+	flags = enter_critical_section();
+	wdog = domain_timer_map[domain_id];
+	/* If delay is zero then cancel the timer (PM Policy) */
+	if (delay == 0) {
+		if (wdog) {
+			timer_timeout(NULL, domain_id);
+		}
```

Comment:
Why when delay == 0, assume timer is already created?. can we check the negative case at start and exit if delay is 0 & no timer available?

Reply:
There is PM Policy to cancel the domain timedsuspend timer if delay is 0.

Delay can't be negative since `milliseconds` argument is of type `uint`.

Reply:
negative case to check, which i mentioned is delay 0 & no timer  (!wdog) at the start.

Reply:
In that case we simply exiting from the code `line:128`

Reply:
If we check delay 0 & no timer (!wdog) at start, we need to again check delay==0 for the delay & timer (wdog) case. So, I think nested if-clause is better for less comparisions